File: app/Models/EnforcementCase.php
<?php

namespace App\Models;

use App\Enums\AppealStatus;
use App\Enums\EnforcementAction;
use App\Enums\EnforcementCaseStatus;
use App\Enums\ReportCategory;
use App\Enums\ViolationSeverity;
use Database\Factories\EnforcementCaseFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

#[Fillable(['user_id', 'related_job_id', 'related_report_id', 'violation_category', 'severity', 'status', 'action', 'restricted_capabilities', 'starts_at', 'ends_at', 'handled_by', 'resolution', 'appeal_status', 'appeal_reason'])]
class EnforcementCase extends Model
{
    /** @use HasFactory<EnforcementCaseFactory> */
    use HasFactory;

    protected function casts(): array
    {
        return ['violation_category' => ReportCategory::class, 'severity' => ViolationSeverity::class, 'status' => EnforcementCaseStatus::class, 'action' => EnforcementAction::class, 'restricted_capabilities' => 'array', 'starts_at' => 'datetime', 'ends_at' => 'datetime', 'appeal_status' => AppealStatus::class];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function relatedJob(): BelongsTo
    {
        return $this->belongsTo(Job::class, 'related_job_id');
    }

    public function relatedReport(): BelongsTo
    {
        return $this->belongsTo(UserReport::class, 'related_report_id');
    }

    public function handler(): BelongsTo
    {
        return $this->belongsTo(User::class, 'handled_by');
    }
}

File: mobile/oncall_mobile/lib/models/enforcement_case.dart
import '../core/json.dart';

class EnforcementCase {
  EnforcementCase({
    required this.id,
    required this.violationCategory,
    required this.severity,
    required this.status,
    this.action,
    required this.restrictedCapabilities,
    this.startsAt,
    this.endsAt,
    this.resolution,
    required this.appealStatus,
    this.appealReason,
    this.createdAt,
  });

  factory EnforcementCase.fromJson(Map<String, dynamic> json) =>
      EnforcementCase(
        id: json['id'] as int,
        violationCategory: json['violation_category'] as String,
        severity: json['severity'] as String,
        status: json['status'] as String,
        action: json['action'] as String?,
        restrictedCapabilities:
            (json['restricted_capabilities'] as List<dynamic>? ?? [])
                .map((e) => e.toString())
                .toList(),
        startsAt: parseDate(json['starts_at']),
        endsAt: parseDate(json['ends_at']),
        resolution: json['resolution'] as String?,
        appealStatus: json['appeal_status'] as String,
        appealReason: json['appeal_reason'] as String?,
        createdAt: parseDate(json['created_at']),
      );

  final int id;
  final String violationCategory;
  final String severity;
  final String status;
  final String? action;
  final List<String> restrictedCapabilities;
  final DateTime? startsAt;
  final DateTime? endsAt;
  final String? resolution;
  final String appealStatus;
  final String? appealReason;
  final DateTime? createdAt;

  bool get canAppeal => appealStatus == 'NONE' && status != 'RESOLVED';
}
